### PR TITLE
feat(create-prs-to-update-vcs-repositories): new github action

### PR DESCRIPTION
**Pull request by sasakisasaki**
## Description

Add a workflow that creates PRs if the expected format tag is attached for vcs imported repositories. 

## Related links

[Semantic Versioning Discussion](https://github.com/orgs/autowarefoundation/discussions/4958)

#### Premise
- Target repositories must have tags matching the pattern: `semantic_version_pattern = r'\b(v?\d+\.\d+(?:\.\d+)?(?:-\w+)?(?:\+\w+(\.\d+)?)?)\b'`
  - This pattern can handle following example tags
```
    "0.0.1",
    "v0.0.1",
    "ros2-v0.0.4",
    "xxx-1.0.0-yyy",
    "2.3.4",
    "v1.2.3-beta",
    "v1.0",
    "v2",
    "1.0.0-alpha+001",
    "v1.0.0-rc1+build.1",
    "2.0.0+build.1848",
    "2.0.1-alpha.1227",
    "1.0.0-alpha.beta",
    "ros_humble-v0.10.2"
```
- If updates occur in multiple repositories, create a separate PR for each repository

#### Test Cases
The following tests passed:

1. **Duplicate PR Prevention**
   - Executing the same Github Action multiple times should not create duplicate PRs with identical content

2. **Multiple Repository Updates**
   - When updates occur simultaneously in two or more repositories, PRs should be created for each affected repository

3. **Idempotent Behavior**
   - The operation should behave as expected and be idempotent, even if the target branch already exists locally or remotely
     - Examples:
         - a. No branches or PRs should be created for anything other than version updates of sub-repositories where tag updates have occurred
         - b. The Github Action should not produce errors like "branch already exists"
         - c. Github Action log output should remain consistent across runs, differing only in timestamps (excluding the initial PR creation)

4. **Post-Merge Behavior**
   - After a PR is merged, no unexpected behavior should occur
     - Examples:
         - a. PRs unrelated to the update should not be created
         - b. If a branch is deleted after the merge process, the deleted branch should not be pushed again

## Notes for reviewers

Perhaps some effective tests are missing in the list of tests above. I'm welcome doing the proposed tests. Your feedback is highly appreciated.

## Interface changes

No change exists in this my PR

## Effects on system behavior

Once this workflow is introduced into the autoware repository, new PRs related to versioning will be created automatically.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by xmfcx** [comment]
What about renaming the workflow?
- from: `create-prs-for-vcs-repositories-update`
- to: `create-prs-to-update-vcs-repositories`

File: create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    # Get GitHub token
+    github_token: str = os.getenv("GITHUB_TOKEN", default=None)
+    if github_token == "None":
+        raise ValueError("Please set GITHUB_TOKEN as an environment variable")
+    github_interface = GitHubInterface(token = github_token)    # cspell: ignore Github
+
+    autoware_repos: AutowareRepos = AutowareRepos(autoware_repos_file_name = args.autoware_repos_file_name)
+
+    # Get the repositories with semantic version tags
+    repository_url_semantic_version_dict: dict[str, str] = autoware_repos.pickup_semver_respositories(semantic_version_pattern = args.semantic_version_pattern)
```
Comment (xmfcx) [inline create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py:226]: typo with respositories

Comment (Shin-kyoto) [inline create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py:226]:
```suggestion
    repository_url_semantic_version_dict: dict[str, str] = autoware_repos.pickup_semver_ repositories(semantic_version_pattern = args.semantic_version_pattern)
```

File: create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+# Initialize logger depending on the verbosity level
+if args.verbose == 0:
+    logging.basicConfig(level=logging.WARNING)
+elif args.verbose == 1:
+    logging.basicConfig(level=logging.INFO)
+elif args.verbose >= 2:
+    logging.basicConfig(level=logging.DEBUG)
+
+
+logger = logging.getLogger(__name__)
```
Comment (Shin-kyoto) [inline create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py:67]: Please move them after `if __name__ == "__main__"`.

Can you make `parse_args` function?
https://github.com/tier4/autoware-ml/blob/28cd951e5d7993896aff0c8c053908a799b47edb/tools/detection3d/visualize_bev.py#L94

**Review by Shin-kyoto — CHANGES_REQUESTED** [review]

File: create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py (hunk 1 above)
Comment (sasakisasaki) [reply create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py]: Thank you! Now fixed by [this commit](https://github.com/autowarefoundation/autoware-github-actions/pull/317/commits/4c82d6aff1f3d811ef53cb550fb6bcfae9dfee7f).

Comment (sasakisasaki) [reply create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py]: Thank you! I also fixed the function definition side too by [this commit](https://github.com/autowarefoundation/autoware-github-actions/pull/317/commits/4c82d6aff1f3d811ef53cb550fb6bcfae9dfee7f).

File: create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py (hunk 2 above)
Comment (sasakisasaki) [reply create-prs-for-vcs-repositories-update/create_prs_for_vcs_repositories_update.py]: I appreciate your proposal. Now applied by [this commit](https://github.com/autowarefoundation/autoware-github-actions/pull/317/commits/f7546c68c8aad92193a3b093d0b589db7c9e60f1).

**Comment by sasakisasaki** [comment]
@xmfcx Thank you so much for your review!

> What about renaming the workflow?
> 
> * from: `create-prs-for-vcs-repositories-update`
> * to: `create-prs-to-update-vcs-repositories`

Good to me! Renamed :+1: .

**Comment by sasakisasaki** [comment]
@Shin-kyoto Thank you so much for your review! Now I applied the suggestions :+1: .

**Comment by sasakisasaki** [comment]
Now all checks looks good :eyes:

**Review by Shin-kyoto — APPROVED** [review]
Thanks!! LGTM.

**Pull request merged**